### ISSUE TITLE
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 182: ordinal not in range(128)

### ISSUE THREAD

**Issue by n8felton** [autopkg/autopkg#111]
```
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 1365, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 1359, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 1146, in run_recipes
    recipe = load_recipe(recipe_path, override_dirs, search_dirs)
  File "/usr/local/bin/autopkg", line 269, in load_recipe
    recipe = load_recipe(parent_id, [], recipe_dirs)
  File "/usr/local/bin/autopkg", line 251, in load_recipe
    name, override_dirs + recipe_dirs)
  File "/usr/local/bin/autopkg", line 170, in find_recipe_by_identifier
    if get_identifer_from_recipe_file(match) == identifier:
  File "/usr/local/bin/autopkg", line 154, in get_identifer_from_recipe_file
    recipe_plist = read_recipe_plist(filename)
  File "/usr/local/bin/autopkg", line 118, in read_recipe_plist
    log_err("WARNING: plist error for %s: %s" % (filename, unicode(err)))
  File "/usr/local/bin/autopkg", line 54, in log_err
    log(msg, error=True)
  File "/usr/local/bin/autopkg", line 47, in log
    print >> sys.stderr, msg
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 182: ordinal not in range(128)
```

**Comment by tincansandtwine** [comment]
Hey, just wanted to comment on this; I ran into a similar issue but with a different line reference. The issue was with an override that I had messed up while manually configuring. Went back and checked, and I had pasted text between a key and a string.

**Comment by gerardweese** [comment]
thanks for the update @tincansandtwine reviewing and correcting a poorly written override(mine) corrected this issue for me.

**Comment by dayglojesus** [comment]
I am seeing this in version 0.4.2 when autopkg encounters a recipe that has some chars that not URL-encoded. In my case it was a bunch of angle brackets [`<` and `>`] inside a `postinstall_script` tag. 

You need to URL encode stuff like this. Not really a problem with autopkg perse.

**Comment by n8felton** [comment]
So I've been able to reliably make this error occur due to any improperly formatted plist in the `RECIPE_SEARCH_DIRS` or `RECIPE_OVERRIDE_DIRS`

Example:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Identifier</key>
    <string>com.github.n8felton.download.UnicodeError</string>
	<key>Input</key>
    <dict>
        <key>NAME</key>
        <string>UnicodeError</string>
    </dict>
	<key>Process</key>
</dict>
</plist>
```

Notice that the `Process` key is missing a value. This can be confirmed with `plutil -lint UnicodeError.munki.recipe`

```
$ plutil -lint UnicodeError.munki.recipe
UnicodeError.munki.recipe: Value missing for key inside <dict> at line 13
```

From what I can tell (with my limited knowledge on python, PyObjC, etc.) it looks like FoundationPlist is getting an NSError about the plist being improperly formated, but is having a hard time handling the error? Looks like we need either better plist validation, and/or better error handling.

**Comment by n8felton** [comment]
I did a little more digging and found this: https://mikeash.com/pyblog/friday-qa-2009-11-20-probing-cocoa-with-pyobjc.html

In the errors section, he mentions adding `.description().encode('utf8')` to the end of the `error` object returned from `NSPropertyListSerialization.propertyListWithData_options_format_error_`

Following https://developer.apple.com/library/prerelease/mac/documentation/Cocoa/Reference/Foundation/Classes/NSError_Class/index.html I did `error.localizedDescription().encode('utf8')` and was able to at least get some basic testing code to work, but it's still not 100%. Leaving this here in case someone can do better with it.

##### FoundationPlist.py
```
def _dataToPlist(data):
    '''low-level function that parses a data object into a propertyList object'''
    darwin_vers = int(os.uname()[2].split('.')[0])
    if darwin_vers > 10:
        (plistObject, plistFormat, error) = (
            NSPropertyListSerialization.propertyListWithData_options_format_error_(
                data, NSPropertyListMutableContainersAndLeaves, None, None))
    else:
        # 10.5 doesn't support propertyListWithData:options:format:error:
        # 10.6's PyObjC wrapper for propertyListWithData:options:format:error:
        #        is broken
        # so use the older NSPropertyListSerialization function
        (plistObject, plistFormat, error) = (
            NSPropertyListSerialization.propertyListFromData_mutabilityOption_format_errorDescription_(
                         data, NSPropertyListMutableContainersAndLeaves, None, None))
    if plistObject is None:
        if error is None:
            error = "Plist data is invalid and could not be deserialized."
        print error.localizedDescription().encode('utf8')
        print error.userInfo().get('NSDebugDescription')
        # raise NSPropertyListSerializationException(error)
    else:
        return plistObject
```

##### Simple test code
```
#!/usr/bin/python
import FoundationPlist

FoundationPlist.readPlist('UnicodeError.munki.recipe')
```

##### Results
```
The data couldn’t be read because it isn’t in the correct format.
Value missing for key inside <dict> at line 13
```

If I try just changing `raise NSPropertyListSerializationException(error)` to `raise NSPropertyListSerializationException(error.localizedDescription().encode('utf8'))` I start getting a new error:
```
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 1662, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 1656, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 925, in get_info
    if get_recipe_info(arguments[0], override_dirs, search_dirs):
  File "/usr/local/bin/autopkg", line 396, in get_recipe_info
    recipe = load_recipe(recipe_name, override_dirs, recipe_dirs)
  File "/usr/local/bin/autopkg", line 332, in load_recipe
    search_github=search_github)
  File "/usr/local/bin/autopkg", line 231, in locate_recipe
    if valid_recipe_file(name):
  File "/usr/local/bin/autopkg", line 147, in valid_recipe_file
    recipe_plist = recipe_plist_from_file(filename)
  File "/usr/local/bin/autopkg", line 110, in recipe_plist_from_file
    "WARNING: plist error for %s: %s" % (filename, unicode(err)))
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 15: ordinal not in range(128)
```

**Comment by gregneagle** [comment]
@n8felton: I created a purposely_broken.recipe with the contents you list above and saved it in ~/Library/AutoPkg/RecipeOverrides/.

I don't get your unhanded exception. I get:
```
% autopkg info Firefox.download.recipe
WARNING: plist error for /Users/gneagle/Library/AutoPkg/RecipeOverrides/intentionally_broken.recipe: Error Domain=NSCocoaErrorDomain Code=3840 "Value missing for key inside <dict> at line 13" UserInfo={NSDebugDescription=Value missing for key inside <dict> at line 13, kCFPropertyListOldStyleParsingError=Error Domain=NSCocoaErrorDomain Code=3840 "Malformed data byte group at line 1; invalid hex" UserInfo={NSDebugDescription=Malformed data byte group at line 1; invalid hex}}
Description:         Download recipe for Firefox. Finds and downloads a Firefox release.
                     Some useful values for RELEASE are: 'latest', 'esr-latest', 'beta-latest'.
<snip>
```

Help me reproduce your error...

**Comment by n8felton** [comment]
Here are the steps I used to reproduce on a vanilla box.
```
sw_vers
ProductName:    Mac OS X
ProductVersion: 10.10.5
BuildVersion:   14F1605
```

###### Install latest AutoPKG
```
cd ~/Downloads
AUTOPKG_LATEST=$(curl https://api.github.com/repos/autopkg/autopkg/releases | python -c 'import json,sys;obj=json.load(sys.stdin);print obj[0]["assets"][0]["browser_download_url"]')
curl -LO# "${AUTOPKG_LATEST}"
sudo installer -pkg $(basename "${AUTOPKG_LATEST}") -target /
```

```
autopkg version
0.5.2
```

###### Install the latest Xcode command line tools.
```
curl https://raw.githubusercontent.com/timsutton/osx-vm-templates/master/scripts/xcode-cli-tools.sh | bash
```

```
autopkg repo-add recipes
autopkg make-override GoogleChrome.munki #"Cheat" to make Override directory
cd ~/Library/AutoPkg/RecipeOverrides/
```
```
$ vi purposely_broken.recipe
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>Identifier</key>
    <string>com.github.n8felton.download.UnicodeError</string>
    <key>Input</key>
    <dict>
        <key>NAME</key>
        <string>UnicodeError</string>
    </dict>
    <key>Process</key>
</dict>
</plist>
```

```
$ plutil -lint purposely_broken.recipe
purposely_broken.recipe: Value missing for key inside <dict> at line 13
```
```
$ autopkg info purposely_broken.recipe
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 1662, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 1656, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 925, in get_info
    if get_recipe_info(arguments[0], override_dirs, search_dirs):
  File "/usr/local/bin/autopkg", line 396, in get_recipe_info
    recipe = load_recipe(recipe_name, override_dirs, recipe_dirs)
  File "/usr/local/bin/autopkg", line 332, in load_recipe
    search_github=search_github)
  File "/usr/local/bin/autopkg", line 236, in locate_recipe
    recipe_file = find_recipe(name, override_dirs + recipe_dirs)
  File "/usr/local/bin/autopkg", line 191, in find_recipe
    return (find_recipe_by_identifier(id_or_name, search_dirs) or
  File "/Library/AutoPkg/autopkglib/__init__.py", line 141, in find_recipe_by_identifier
    if get_identifier_from_recipe_file(match) == identifier:
  File "/Library/AutoPkg/autopkglib/__init__.py", line 124, in get_identifier_from_recipe_file
    "WARNING: plist error for %s: %s" % (filename, unicode(err)))
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2019' in position 156: ordinal not in range(128)
```

**Comment by gregneagle** [comment]
I'm guessing the most relevant difference is that I'm running 10.11.3, and the error message generated under that OS doesn't seem to have non-ASCII characters.

**Comment by zackbennis** [comment]
I got this error continuously using Autopkgr after modifying some overrides. I went back and triple-checked (or more) all my recipes, made sure everything was encoded properly, and still got the issue. It was only when running autopkg manually in the Terminal that I think I found the culprit:

```
WARNING: plist error for /Users/username/Library/AutoPkg/NewOverrides/MSWord2016.munki.recipe: Error Domain=NSCocoaErrorDomain Code=3840 "The data couldn’t be read because it isn’t in the correct format." (Found non-key inside <dict> at line 30) UserInfo=0x7f94d142aec0 {NSDebugDescription=Found non-key inside <dict> at line 30, kCFPropertyListOldStyleParsingError=The data couldn’t be read because it isn’t in the correct format.}
```

Note the right handed quote, i.e. u2019, in the error message. I believe this is what has been causing me great heartache.

**Comment by gregneagle** [comment]
And since we (or at least I) cannot reliably reproduce the issue, it's been nearly impossible to fix.  I broke my MSWord2016.munki recipe override and I get this error instead:

```
% autopkg run MSWord2016.munki.recipe
WARNING: plist error for /Users/gneagle/Library/AutoPkg/RecipeOverrides/MSWord2016.munki.recipe: Error Domain=NSCocoaErrorDomain Code=3840 "Found non-key inside <dict> at line 30" UserInfo={NSDebugDescription=Found non-key inside <dict> at line 30, kCFPropertyListOldStyleParsingError=Error Domain=NSCocoaErrorDomain Code=3840 "Malformed data byte group at line 1; invalid hex" UserInfo={NSDebugDescription=Malformed data byte group at line 1; invalid hex}}
WARNING: plist error for /Users/gneagle/Library/AutoPkg/RecipeOverrides/MSWord2016.munki.recipe: Error Domain=NSCocoaErrorDomain Code=3840 "Found non-key inside <dict> at line 30" UserInfo={NSDebugDescription=Found non-key inside <dict> at line 30, kCFPropertyListOldStyleParsingError=Error Domain=NSCocoaErrorDomain Code=3840 "Malformed data byte group at line 1; invalid hex" UserInfo={NSDebugDescription=Malformed data byte group at line 1; invalid hex}}
Processing MSWord2016.munki.recipe...
```

Of course, I had to guess at what to do to "break" the recipe. Until we have a consistent way to reproduce this issue, we're not going to make much progress.

**Comment by gregneagle** [comment]
Testing on a 10.9.5 machine, I do get a curly quote in the error message:

```
$ autopkg run /var/madmin/Library/AutoPkg/RecipeOverrides/Skype.download.recipe
WARNING: plist error for /var/madmin/Library/AutoPkg/RecipeOverrides/Skype.download.recipe: Error Domain=NSCocoaErrorDomain Code=3840 "The data couldn’t be read because it isn’t in the correct format." (Found non-key inside <dict> at line 10) UserInfo=0x7fd5a8e0e7a0 {NSDebugDescription=Found non-key inside <dict> at line 10, kCFPropertyListOldStyleParsingError=The data couldn’t be read because it isn’t in the correct format.}
```

So is this _not_ an autopkg issue, but actually an _AutoPkgr_ issue?

**Comment by zackbennis** [comment]
As best I understand the issue: an error that autopkg can handle just fine produces a useful warning that contains an unsupported character and the culprit file. The curly quote _**in the error message that autopkg produces**_ is causing AutoPkgr to fail in only 1 way, with the not-so-useful UnicodeEncodeError/u2019 that doesn't single out the file causing the warning in autopkg.

Seems like it might be an AutoPkgr problem, but I don't know for sure; is the error message being generated by autopkg? If so, is it possible to change it to a supported character?

**Comment by gregneagle** [comment]
No, autopkg is just reporting an error generated by Apple code when trying to read a plist. This is why the exact error differs somewhat depending on OS X version.

**Comment by macmule** [comment]
Hi folks,

I think I have a reproducible method for this.

The below is a snippet from a recipe override that I did a copy/paste fail on:

```
		<key>MUNKI_REPO_SUBDIR</key>
		<string>%CATEGORY%/%DEVELOPER%/%NAME%</string>
		<string>%CATEGORY%/%DEVELOPER%/%NAME%</string>
```

Plutil picks this up: `Found non-key inside <dict> at line 11`

But AutoPKGr errors as per:
![screenshot 2017-01-16 11 36 50](https://cloud.githubusercontent.com/assets/2464974/21981682/7f63645a-dbe0-11e6-9037-0fb2cf3923af.png)

autopkg run -vvv on the recipe gives: 
```
autopkg run -vvv ~/Library/AutoPkg/RecipeOverrides/jHelper\ GUI-dataJAR.munki.recipe
WARNING: plist error for /Users/sadmin/Library/AutoPkg/RecipeOverrides/jHelper GUI-dataJAR.munki.recipe: Error Domain=NSCocoaErrorDomain Code=3840 "The data couldn’t be read because it isn’t in the correct format." (Found non-key inside <dict> at line 11) UserInfo=0x7ff218f8dbf0 {NSDebugDescription=Found non-key inside <dict> at line 11, kCFPropertyListOldStyleParsingError=The data couldn’t be read because it isn’t in the correct format.}
```

So.. I'd be looking to AutoPKGr's handling of the error

**Comment by gregneagle** [comment]
so shouldn't this be an issue for autopkgR, then?

**Comment by macmule** [comment]
Logged as AutoPKGr issue. Think this can be closed off here.

**Comment by n8felton** [comment]
Found another way to generate this error.

1. Install Xcode
2. Do _not_ run `xcodebuild -license accept` 
3. `autopkg repo-add http://github.com/autopkg/recipes` causes the error message:

```
Attempting git clone...
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 2435, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 2429, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 702, in repo_add
    new_recipe_repo_dir = get_recipe_repo(repo_url)
  File "/usr/local/bin/autopkg", line 551, in get_recipe_repo
    log_err(err)
  File "/usr/local/bin/autopkg", line 73, in log_err
    log(msg, error=True)
  File "/usr/local/bin/autopkg", line 66, in log
    print >> sys.stderr, unicode(msg).encode('UTF-8')
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 81: ordinal not in range(128)
```
4. Manually run `git clone http://github.com/autopkg/recipes` and receive the message:

`Agreeing to the Xcode/iOS license requires admin privileges, please run “sudo xcodebuild -license” and then retry this command.`

Notice the "smart" quotes in the message?

**Comment by mikep345678** [comment]
For several weeks I've had a number of autopkg repo-updates fail with this UnicodeDecodeError.

I'd seen this issue thread but it doesn't (overtly) present a solution or come to a conclusion, so yesterday I finally had a chance to hit the Slack to see if anyone else had similar experiences or any solutions to suggest.

It was determined that some of the repo-updates (and/or repo-adds) that failed for me, _didn't_ fail for other folks. This helped to point me that this was a config issue on my machine.

We understood that this UnicodeDecodeError is sort of a meta-error, an error about an error, and there were several ideas to check to try to fix it-- wrong Python version issues, poorly-written recipes, corrupt autopkg settings. Troubleshooting through those hints didn't resolved the issue for me, though.

Rescouring this thread, I finally understood the implications of @n8felton's most recent posting about xcodebuild's smart-quotes, tried re-accepting the license, and-- viola, the UnicodeDecodeError's went away.

tl;dr: running `“sudo xcodebuild -license accept”` resolved the issue for me.


Mike

**Comment by gregneagle** [comment]
Is there a way to "unaccept" the Xcode license? It's really hard to troubleshoot and fix an error you cannot reproduce.

**Comment by rtrouton** [comment]
I just ran into this issue as well. Here's the output:

```
computername:~ username$ autopkg repo-add apettinen-recipes
Attempting git clone...
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 2453, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 2447, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 703, in repo_add
    new_recipe_repo_dir = get_recipe_repo(repo_url)
  File "/usr/local/bin/autopkg", line 552, in get_recipe_repo
    log_err(err)
  File "/usr/local/bin/autopkg", line 73, in log_err
    log(msg, error=True)
  File "/usr/local/bin/autopkg", line 66, in log
    print >> sys.stderr, unicode(msg).encode('UTF-8')
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 81: ordinal not in range(128)
computername:~ username$ autopkg repo-add arubdesu-recipes
Attempting git clone...
Traceback (most recent call last):
  File "/usr/local/bin/autopkg", line 2453, in <module>
    sys.exit(main(sys.argv))
  File "/usr/local/bin/autopkg", line 2447, in main
    exit(subcommands[verb]['function'](argv))
  File "/usr/local/bin/autopkg", line 703, in repo_add
    new_recipe_repo_dir = get_recipe_repo(repo_url)
  File "/usr/local/bin/autopkg", line 552, in get_recipe_repo
    log_err(err)
  File "/usr/local/bin/autopkg", line 73, in log_err
    log(msg, error=True)
  File "/usr/local/bin/autopkg", line 66, in log
    print >> sys.stderr, unicode(msg).encode('UTF-8')
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 81: ordinal not in range(128)
computername:~ username$ git clone http://github.com/autopkg/recipes


Agreeing to the Xcode/iOS license requires admin privileges, please run “sudo xcodebuild -license” and then retry this command.


computername:~ username$ sudo xcodebuild -license
Password:

<skip past license stuff>
<agree to license>

username:~ username$ git clone http://github.com/autopkg/recipes
Cloning into 'recipes'...
warning: redirecting to https://github.com/autopkg/recipes/
remote: Enumerating objects: 4, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (4/4), done.
remote: Total 3030 (delta 0), reused 0 (delta 0), pack-reused 3026
Receiving objects: 100% (3030/3030), 710.01 KiB | 10.60 MiB/s, done.
Resolving deltas: 100% (2016/2016), done.
username:~ username$ autopkg repo-add apettinen-recipes
Attempting git clone...

Adding /Users/username/Library/AutoPkg/RecipeRepos/com.github.autopkg.apettinen-recipes to RECIPE_SEARCH_DIRS...
Updated search path:
  '.'
  '~/Library/AutoPkg/Recipes'
  '/Library/AutoPkg/Recipes'
  '/Users/username/Library/AutoPkg/RecipeRepos/com.github.autopkg.apettinen-recipes'
computername:~ username$ autopkg repo-add arubdesu-recipes
Attempting git clone...

Adding /Users/username/Library/AutoPkg/RecipeRepos/com.github.autopkg.arubdesu-recipes to RECIPE_SEARCH_DIRS...
Updated search path:
  '.'
  '~/Library/AutoPkg/Recipes'
  '/Library/AutoPkg/Recipes'
  '/Users/username/Library/AutoPkg/RecipeRepos/com.github.autopkg.apettinen-recipes'
  '/Users/username/Library/AutoPkg/RecipeRepos/com.github.autopkg.arubdesu-recipes'
computername:~ username$
```
